File: books/sword_offer/_0_find_Kth_to_tail.java

public class _0_find_Kth_to_tail {
    public static void main(String[] args) {
    }

    public FindKthToTail(ListNode pListHead, int k) {
        if (pListHead == null) {
            return null;
        }

        ListNode pAhead = pListHead;
        ListNode pBhead = null;

        for (int i=0; i<k; i++) {
            if (pAhead == null) {
                return null;
            }
            pAhead = pAhead.m_pNext;
        }

        pBhead = pListHead;

        while (pAhead != null) {
            pAhead = pAhead.m_pNext;
            pBhead = pBhead.m_pNext;
        }
        return pBhead;

    }
}

class ListNode {
    public int val;
    public ListNode m_pNext;

    public ListNode(int val, ListNode m_pNext) {
        this.val = val;
        this.m_pNext = m_pNext;
    }

    public ListNode() {
        this(0, null);
    }
}

File: books/sword_offer/0_find_Kth_to_tail.go
package main

type _ListNode struct {
	val			int
	m_pNext		*_ListNode
}

/**
	desc：	求链表中的倒数第k个节点

	tips：	对象的创建、指针
 */


func (self *_ListNode) newListNode() *_ListNode {
	return &_ListNode{val:0, m_pNext:nil}
}

func FindKthToTail(pListHead *_ListNode, k int) *_ListNode {
	if pListHead == nil {
		return nil
	}

	pHead := pListHead
	var pBehind *_ListNode = pListHead

	for i:=0; i<k-1;  {
		if pHead == nil {
			return nil
		}
		pHead = pHead.m_pNext
	}

	for pHead != nil {
		pHead = pHead.m_pNext
		pBehind = pBehind.m_pNext
	}
	return pBehind
}



